### PR TITLE
Gene ontology(go) adapter

### PR DESCRIPTION
**Pull request by kedistS**
This is an adapter for gene ontology(go).

### CODE REVIEW

File: biocypher_metta/adapters/gene_ontology_adapter.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    def __init__(self, write_properties, add_provenance, type, label='ontology term', dry_run=False):
+        super().__init__(write_properties, add_provenance, label, type, dry_run)
+        self.source = "Gene Ontology"
+        self.source_url = "http://purl.obolibrary.org/obo/go.owl"
+        self.ONTOLOGIES = {
+            'go': 'http://purl.obolibrary.org/obo/go.owl'
+        }
+
+
+    def get_graph(self, ontology='go'):
```
Comment (Habush) [inline biocypher_metta/adapters/gene_ontology_adapter.py]: This method already exists in the parent class and the one in the parent class should be used

File: biocypher_metta/adapters/gene_ontology_adapter.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        if ontology not in self.ONTOLOGIES:
+            raise ValueError(f"Ontology '{ontology}' is not defined in this adapter.")
+        
+        onto = get_ontology(self.ONTOLOGIES[ontology]).load()
+        self.graph = default_world.as_rdflib_graph()
+        self.clear_cache()
+        return self.graph
+
+    def get_nodes(self):
```
Comment (Habush) [inline biocypher_metta/adapters/gene_ontology_adapter.py:38]: What's the difference between the `get_nodes` and `get_edges` methods in this class and in the `OntologyAdapter` parent class? I don't see much of a difference. If there is no difference, perhaps no need to override it here.

File: config/adapters_config.yaml [hunk 3]
```diff
@@ -137,30 +137,32 @@ child_pathway_of:
...
   adapter:
-    module: biocypher_metta.adapters.ontologies_adapter
-    cls: OntologyAdapter
+    module: biocypher_metta.adapters.gene_ontology_adapter
+    cls: GeneOntologyAdapter
     args:
+      label: subtype_of
       type: edge
-      dry_run: False
+      dry_run: True
```
Comment (Habush) [inline config/adapters_config.yaml]: `dry_run` should be `False` in the main adapters config. It should be set `True` in the sample adapters config.

**Review by Habush — CHANGES_REQUESTED** [review]

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 4]
```diff
@@ -1,168 +1,143 @@
...
+        if ontology not in self.ONTOLOGIES:
+            raise ValueError(f"Ontology '{ontology}' is not defined in this adapter.")
+    
+        onto = get_ontology(self.ONTOLOGIES[ontology]).load()
+        self.graph = default_world.as_rdflib_graph()  # Assign the graph to self.graph
         self.clear_cache()
         return self.graph
 
     def get_nodes(self):
+        self.graph = self.get_graph()
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: `get_graph`  is being called without the `ontology`  argument

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 5]
```diff
@@ -1,168 +1,143 @@
...
-            self.graph = self.get_graph(ontology)
-            self.cache_node_properties()
-
-            nodes_in_go_namespaces = self.find_go_nodes(self.graph)
-            nodes = nodes_in_go_namespaces.keys()
-
-            i = 0 # dry run is set to true just output the first 1000 nodes
-            for node in nodes:
+    def get_edges(self):
+        self.graph = self.get_graph()
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: `get_graph` is being called without the `ontology` argument

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 6]
```diff
@@ -192,7 +164,7 @@ def to_key(cls, node_uri):
             key = '{}_{}'.format('number', key)
 
         return key
-
+
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py:174]: shouldn't there be a method definition here?

File: biocypher_metta/adapters/gene_ontology_adapter.py (hunk 1 above)
Comment (Habush) [reply biocypher_metta/adapters/gene_ontology_adapter.py]: I still don't see a reason for this method to exist here. can't you call `self.graph(ontology='go')` when you want the `graph` method in the parent class

**Review by Habush — CHANGES_REQUESTED** [review]

File: biocypher_metta/adapters/ontologies_adapter.py (hunk 6 above)
Comment (kedistS) [reply biocypher_metta/adapters/ontologies_adapter.py:174]: Could you please clarify which method you are referring to?

File: biocypher_metta/adapters/gene_ontology_adapter.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+from biocypher_metta.adapters.ontologies_adapter import OntologyAdapter
+
+class GeneOntologyAdapter(OntologyAdapter):
+    def __init__(self, write_properties, add_provenance, ontology, type, label="ontology term", dry_run=False):
```
Comment (Habush) [inline biocypher_metta/adapters/gene_ontology_adapter.py]: As per the schema, the label name should be `go`

File: biocypher_metta/adapters/gene_ontology_adapter.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        super().__init__(write_properties, add_provenance, ontology, type, label, dry_run)
```
Comment (Habush) [inline biocypher_metta/adapters/gene_ontology_adapter.py]: Minor but better to pass arguments to the parent class after instantiating the fields of the current class

File: biocypher_metta/adapters/gene_ontology_adapter.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        self.source = "Gene Ontology"
+        self.source_url = "http://purl.obolibrary.org/obo/go.owl"
+        self.ontology = ontology
+        self.ONTOLOGIES = {
```
Comment (Habush) [inline biocypher_metta/adapters/gene_ontology_adapter.py]: Where is this field used?

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 10]
```diff
@@ -1,168 +1,144 @@
...
-    def get_graph(self, ontology):
-
-        onto = get_ontology(OntologyAdapter.ONTOLOGIES[ontology]).load()
+        self.graph = None
+        self.cache = {}
+        self.source = None
+        self.source_url = None
+        self.ontology = ontology  # Store the ontology value
+
+    def get_graph(self):
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: This method sets the `self.graph` field and then returns it. But it's unnecessary to return the field after setting it.

Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: I'd suggest to change the method's name to `update_graph` and then set the `graph` field in it without returning it. Then on the `get_nodes` method, you'd just call `self.update_graph` and the `self.graph` field now contains the updated ontology graph

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 11]
```diff
@@ -1,168 +1,144 @@
...
+            raise ValueError(f"Ontology '{self.ontology}' is not defined in this adapter.")
+    
+        onto = get_ontology(self.ONTOLOGIES[self.ontology]).load()
         self.graph = default_world.as_rdflib_graph()
-        # self.graph = onto.as_rdflib_graph()
         self.clear_cache()
         return self.graph
 
     def get_nodes(self):
+        self.graph = self.get_graph()
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: Change this to `self.update_graph()`.

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 12]
```diff
@@ -1,168 +1,144 @@
...
+    EXACT_SYNONYM = rdflib.term.URIRef('http://www.geneontology.org/formats/oboInOwl#hasExactSynonym')
+    RELATED_SYNONYM = rdflib.term.URIRef('http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym')
+    DESCRIPTION = rdflib.term.URIRef('http://purl.obolibrary.org/obo/IAO_0000115')
 
     PREDICATES = [SUBCLASS, DB_XREF]
     RESTRICTION_PREDICATES = [HAS_PART, PART_OF]
 
-    def __init__(self, type='node', dry_run=False):
+    def __init__(self, write_properties, add_provenance, ontology, type, label, dry_run=False):
+        super().__init__(write_properties, add_provenance)
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: Minor but better to pass arguments to the parent class after instantiating the fields of the current class

File: biocypher_metta/adapters/gene_ontology_adapter.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+            node = n[0]
+            namespace = n[1]
+
+            node_key = OntologyAdapter.to_key(node)
+            node_namespace_lookup[node_key] = str(namespace)
+        return node_namespace_lookup
+
+    def get_nodes(self):
+        nodes = super().get_nodes()
+        self.graph = self.get_graph()
```
Comment (Habush) [inline biocypher_metta/adapters/gene_ontology_adapter.py]: `get_graph` is being called in the parent's `get_nodes` method, so why it being called here again?

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 14]
```diff
@@ -1,168 +1,144 @@
...
-            self.graph = self.get_graph(ontology)
-            self.cache_node_properties()
-
-            nodes_in_go_namespaces = self.find_go_nodes(self.graph)
-            nodes = nodes_in_go_namespaces.keys()
-
-            i = 0 # dry run is set to true just output the first 1000 nodes
-            for node in nodes:
+    def get_edges(self):
+        self.graph = self.get_graph()
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: Change this to `self.update_graph()`.

**Review by Habush — CHANGES_REQUESTED** [review]

File: biocypher_metta/adapters/ontologies_adapter.py [hunk 15]
```diff
@@ -1,168 +1,145 @@
...
+        self.source, self.source_url = self.get_ontology_source(ontology)
+
+        super(OntologyAdapter, self).__init__(write_properties, add_provenance)
+    
+    def get_ontology_source(self, ontology):
+        """
+        Returns the source and source URL for a given ontology.
+        This method should be overridden in child classes for specific ontologies.
+        """
+        return None, None
```
Comment (Habush) [inline biocypher_metta/adapters/ontologies_adapter.py]: To indicate this method is abstract, just relapce the `return` statement with `pass`.

**Review by Habush — CHANGES_REQUESTED** [review]

**Review by Habush — APPROVED** [review]
Looks good.

**Pull request merged**